extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/QuantilesSketchAggregatorFactory.java

```diff
@@ -0,0 +1,327 @@
+      @JsonProperty("fieldName") String fieldName,
+      @JsonProperty("size") Integer size,
+      @JsonProperty("shouldFinalize") Boolean shouldFinalize,
+      @JsonProperty("isInputSketch") boolean isInputSketch
+  )
+  {
+    this.name = Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    this.fieldName = Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+
+    this.size = size == null ? DEFAULT_MAX_SKETCH_SIZE : size;
+
+    //validate the size with sketch library
+    Util.checkIfPowerOf2(this.size, "size");
+    //dummy sketch creation just to ensure all size restrictions are fulfilled.
+    QuantilesSketchUtils.buildSketch(this.size);
+
+    this.shouldFinalize = (shouldFinalize == null) ? true : shouldFinalize.booleanValue();
+    this.isInputSketch = isInputSketch;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(fieldName);
+    if (selector == null) {
+      return new EmptyQuantilesSketchAggregator(name);
+    } else {
+      return new QuantilesSketchAggregator(name, selector, size);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(fieldName);
+    if (selector == null) {
+      return new EmptyQuantilesSketchBufferAggregator();
+    } else {
+      return new QuantilesSketchBufferAggregator(selector, size, getMaxIntermediateSize());
+    }
+  }
+
+  @Override
+  public Object deserialize(Object object)
+  {
+    return QuantilesSketchUtils.deserialize(object);
+  }
+
+  @Override
+  public Comparator<QuantilesSketch> getComparator()
+  {
+    return COMPARATOR;
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    if (lhs == null || ((QuantilesSketch) lhs).isEmpty()) {
+      return rhs;
+    }
+
+    if (rhs == null || ((QuantilesSketch) rhs).isEmpty()) {
+      return lhs;
+    }
+
+    Union union = QuantilesSketchUtils.buildUnion(size);
+    union.update((QuantilesSketch) lhs);
+    union.update((QuantilesSketch) rhs);
+
+    return union.getResult();
+  }
+
+  @Override
+  @JsonProperty
+  public String getName()
+  {
+    return name;
+  }
+
+  @JsonProperty
+  public String getFieldName()
+  {
+    return fieldName;
+  }
+
+  @JsonProperty
+  public int getSize()
+  {
+    return size;
+  }
+
+  @JsonProperty
+  public boolean getShouldFinalize()
+  {
+    return shouldFinalize;
+  }
+
+  @JsonProperty
+  public boolean getIsInputSketch()
+  {
+    return isInputSketch;
+  }
+
+  @Override
+  public int getMaxIntermediateSize()
+  {
+    // off-heap implementation of quantiles sketch is not available right now and hence
+    // nothing is stored off-heap really.
+    return 1;
```

Comment:
I'm with @gianm on this one.  The implementation here breaks the current model that all other aggregators use and if you want to use it, you now have to tune your heap size differently than you previously had to, and you only have to tune it for this one query.  The "fix" in #2827 will end up enforcing some global limits on how much work can be done on any and every query, even when that limitation doesn't make sense.

The semantics inside of Druid is for aggregators to specify how much memory they use and for query implementations to leverage that.  One of the following things should happen

1. Make the implementation off-heap (continuously serialize and deserialize the object over and over again)
2. Throw an `UnsupportedOperationException` from `factorizeBuffered` and hope that this is still usable in some areas (timeseries queries?)
3. Pull all of this into an external module, where adherence with normal assumptions and expectations of how things work don't have to be adhered to.